linera-chain/src/certificate.rs

```diff
@@ -0,0 +1,307 @@
+    ) -> &Vec<(ValidatorName, Signature)> {
+        let index = self
+            .signatures
+            .binary_search_by(|(name, _)| name.cmp(&signature.0))
+            .unwrap_or_else(std::convert::identity);
+        self.signatures.insert(index, signature);
+        &self.signatures
+    }
+
+    /// Returns whether the validator is among the signatories of this certificate.
+    pub fn is_signed_by(&self, validator_name: &ValidatorName) -> bool {
+        self.signatures
+            .binary_search_by(|(name, _)| name.cmp(validator_name))
+            .is_ok()
+    }
+}
+
+impl<T> CertificateT<T> {
+    /// Returns the certified value's hash.
+    pub fn hash(&self) -> CryptoHash {
+        self.value.hash()
+    }
+
+    /// Verifies the certificate.
+    pub fn check(&self, committee: &Committee) -> Result<(), ChainError> {
+        crate::data_types::check_signatures(self.hash(), self.round, &self.signatures, committee)?;
+        Ok(())
+    }
+}
+
+impl ValidatedBlockCertificate {
+    pub fn requires_blob(&self, blob_id: &BlobId) -> bool {
+        self.executed_block().requires_blob(blob_id)
+    }
+
+    #[cfg(with_testing)]
+    pub fn outgoing_message_count(&self) -> usize {
+        self.executed_block().messages().iter().map(Vec::len).sum()
+    }
+
+    /// Returns the `LiteValue` corresponding to the certified value.
+    pub fn lite_value(&self) -> LiteValue {
+        LiteValue {
+            value_hash: self.value.hash(),
+            chain_id: self.executed_block().block.chain_id,
+        }
+    }
+
+    /// Returns the certificate without the full value.
+    pub fn lite_certificate(&self) -> LiteCertificate<'_> {
+        LiteCertificate {
+            value: self.lite_value(),
+            round: self.round,
+            signatures: Cow::Borrowed(&self.signatures),
+        }
+    }
+
+    /// Returns reference to the `ExecutedBlock` contained in this certificate.
+    pub fn executed_block(&self) -> &ExecutedBlock {
+        self.inner().inner()
+    }
+}
+
+impl From<Certificate> for ValidatedBlockCertificate {
+    fn from(cert: Certificate) -> Self {
+        let signatures = cert.signatures().clone();
+        let hash = cert.value.hash();
+        match cert.value.into_inner() {
+            CertificateValue::ValidatedBlock { executed_block } => Self {
+                value: Hashed::unsafe_new(ValidatedBlock::new(executed_block), hash),
+                round: cert.round,
+                signatures,
+            },
+            _ => panic!("Expected a validated block certificate"),
+        }
+    }
+}
+
+impl ConfirmedBlockCertificate {
+    /// Creates a new `ConfirmedBlockCertificate` from a `ValidatedBlockCertificate`.
+    pub fn from_validated(validated: ValidatedBlockCertificate) -> Self {
+        let ValidatedBlockCertificate {
+            round,
+            value: Hashed {
+                value: validated_block,
+                ..
+            },
+            ..
+        } = validated;
```

Comment:
Maybe simpler (copying the round is cheap):

```suggestion
        let round = validated.round;
        let validated_block = validated.value.into_inner();
```

Reply:
Done in https://github.com/linera-io/linera-protocol/pull/2803/commits/718f2bf2d0a65301bf6b5ca8eef24099a8a3235a